daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilInputContentHandler.scala

```diff
@@ -0,0 +1,180 @@
+import org.apache.daffodil.api.DFDL.DaffodilUnparseErrorSAXException
+import org.apache.daffodil.infoset.IllegalContentWhereEventExpected
+import org.apache.daffodil.infoset.InfosetInputterEventType.EndDocument
+import org.apache.daffodil.infoset.InfosetInputterEventType.EndElement
+import org.apache.daffodil.infoset.InfosetInputterEventType.StartDocument
+import org.apache.daffodil.infoset.InfosetInputterEventType.StartElement
+import org.apache.daffodil.infoset.SAXInfosetInputter
+import org.apache.daffodil.util.Maybe.Nope
+import org.apache.daffodil.util.Maybe.One
+import org.xml.sax.Attributes
+import org.xml.sax.Locator
+
+class DaffodilInputContentHandler(dp: DFDL.DataProcessor, output: DFDL.Output)
+  extends DFDL.DaffodilInputContentHandler {
+  private val inputter = new SAXInfosetInputter(this, dp, output)
+  private var unparseResult: DFDL.UnparseResult = _
+  private var infosetEvent: DFDL.SaxInfosetEvent = _
+  private val characterData = new StringBuilder()
+
+  def setUnparseResult(ur: DFDL.UnparseResult): Unit = {
+    unparseResult = ur
+  }
+
+  def getUnparseResult: DFDL.UnparseResult = unparseResult
+
+  override def setDocumentLocator(locator: Locator): Unit = {
+    // do nothing
+  }
+
+  override def startDocument(): Unit = {
+    infosetEvent = new DFDL.SaxInfosetEvent
+    infosetEvent.eventType = One(StartDocument)
+    sendToInputter
+  }
+
+  override def endDocument(): Unit = {
+    if (infosetEvent == null) {
+      infosetEvent = new DFDL.SaxInfosetEvent
+    }
+    infosetEvent.eventType = One(EndDocument)
+    sendToInputter
+  }
+
+  // TODO not sure what to do with the prefix mappings, technically we don't need them since
+  // we have the attributes with startElement to look up any prefixes we need
+  override def startPrefixMapping(prefix: String, uri: String): Unit = {
+    val pre = if (prefix == "") null else prefix
+    infosetEvent.prefixMapping = One(NamespaceBinding(pre, uri, infosetEvent.prefixMapping
+      .orNull))
+  }
+
+  override def endPrefixMapping(prefix: String): Unit = {
+    // do nothing
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // we need to check if the characters info is all whitespace, if not, it is an error
+    if (characterData.nonEmpty && characterData.toString().trim.nonEmpty) {
+      throw new IllegalContentWhereEventExpected("Non-whitespace characters in complex " +
+        "Element: " + characterData.toString())
+    } else {
+      // reset since it was whitespace only
+      characterData.setLength(0)
+    }
+
+    if (infosetEvent == null) infosetEvent = new DFDL.SaxInfosetEvent
+    else if (!infosetEvent.isEmpty && infosetEvent.localName.isDefined) {
+      // we started another element while we were in the process of building another
+      // this most likely means the first element was complex, either ways, it is ready
+      // for the inputter queue
+      sendToInputter
```

Comment:
> most likely means

I think this definitely means, right? I'm not sure I understand the isEmpty or localName.isDefined checks? can you explain those?